### DISCUSSION TITLE
Giving feedback after total success rate in a timeline

### DISCUSSION THREAD

**Discussion by georginatorok** [jspsych/jsPsych#2834] [Q&A]
Hi all, 

I am fairly new to JavaScript and have some difficulties working out how to provide performance-based feedback after a timeline of multiple trials.

In my experiment, I am using the Serial Reaction Time Mouse plugin to give participants a sequence they need to accurately follow. This is implemented in the “exgame_intro_tap” timeline (see code below).

I want to give feedback on the total number of hits (correct trials) in the sequence right after the SRT sequence is done. I managed to calculate the number of hits with an “on_timeline_finish” function after all 6 trials in the sequence are done and assigned it to a variable called “accuracy_exgame_intro_tap“. Console.log() can successfully display this variable’s value while the task is running. However, I am having trouble accessing and using that value in the next timeline where I want to use the number of hits to give feedback on the participant’s success rate. I tried to add it to the data with jsPsych.data.addProperties() in the on_timeline_finish function, but it doesn’t work. I get the following error:

Uncaught ReferenceError: accuracy is not defined

Could you help me with how to save the success rate after a timeline is finished, so that it can be used in a subsequent feedback trial? I’m probably not saving it in the right place/right way. Is this because the .addProperties() works at the very end of the experiment, so at the time of trying to run the feedback trial, the variable is not yet created? Or something to do with a globally vs locally accessible variables issue? Where and how do I need to save the calculated accuracy value in order for it to be accessible for the feedback trial? 

Best,
Gina

``` javascript
var exgame_intro_tap = {
    timeline: [
      {
        type: jsPsychSerialReactionTimeMouse,
        grid: grid,
        target: jsPsych.timelineVariable('target'), 
        target_image: frogs_small[6],
        pre_target_duration: pre_targ_dur,
        trial_duration: trial_dur,
        grid_square_size: square_size,
        prompt: catchfrog,
        allow_nontarget_responses: true,
        data: {
        task: 'sequence', game: 'example_tapping', phase: 'intro', gamePhase: 'exgamefam',
        color: frogs_small[6].slice(frogs_small[6].search('frog')+5, frogs_small[6].search('small')-1),
        condition: "intro", seqID: 'exgame_intro_tap', trial_no: jsPsych.timelineVariable('trial_no')
        }
      },
  ],
  timeline_variables: [
      { target: ex_game[0], trial_no: 1},
      { target: ex_game[1], trial_no: 2},
      { target: ex_game[2], trial_no: 3},
      { target: ex_game[3], trial_no: 4},
      { target: ex_game[4], trial_no: 5},
      { target: ex_game[5], trial_no: 6}
  ],

  on_timeline_finish: function(data) {
      var correct_count = jsPsych.data.get().filter({seqID: 'exgame_intro_tap', correct: true}).trials;
      accuracy_exgame_intro_tap = correct_count.length;
      console.log('The accuracy of the example game was:', accuracy_exgame_intro_tap, ' out of 6 hits');
      
      jsPsych.data.addProperties({
        accuracy: accuracy_exgame_intro_tap
      });
    },

}

console.log('this is what we saved: ', accuracy);

// provide feedback on sequence trial
var exgame_intro_tap_feedback = {
type: jsPsychInstructions,
pages: [
  '<p style = " font-size:150%"><strong>You caught the frog</strong></p>' + accuracy + 
  '<p style = " font-size:150%"><strong>out of 6 times!</strong></p>' +
  '<img src="img/grid_star'+ accuracy +'.png"></img>' +
  '<p style = " font-size:150%"><strong> Tap on the arrow to move on!</strong></p>' 
],
allow_keys: false,
allow_backward: false,
button_label_next: '>>>' ,
show_clickable_nav: true
};
```

**Comment by jodeleeuw** [comment]
Hi @georginatorok,

The issue that you are running into is because the `exgame_intro_tap_feedback` variable is created before the experiment starts to run. So at the time that it is created, the `accuracy` variable doesn't exist yet.

The way to solve this is with [dynamic parameters](https://www.jspsych.org/7.3/overview/dynamic-parameters/). There's also [a step in the main tutorial](https://www.jspsych.org/7.3/tutorials/rt-task/#part-13-data-aggregation) that shows how to aggregate this kind of data and show it to the user.

Feel free to follow up here if those resources don't solve the problem!

**Reply by georginatorok** [reply]
Hi Josh,
Thanks very much for the suggestions, I managed to get it work! I had kept reading the documentation but just couldn't put two and two together myself... the tutorial example helped immensely. Much appreciated!

**Reply by georginatorok** [reply]
Hi Josh, 
I have another question if you don't mind, since I see that you wrote the SRT plugin - is there a way to change it so that RTs of late clicks are also shown in the data file? Now when a participant taps late for target t=1, but _before_ the t=2 target appears (ie., the grid is empty during the inter-trial interval), I have no RT. I would like to know how (a)synchronous with the sequence people's tapping actions are. 

In the plugin file, to achieve this, would I need to change something in this part? 

``` javascript
// function to handle responses by the subject
          function after_response(info) {
              // only record first response
              response = response.rt == null ? info : response;
              if (trial.response_ends_trial) {
                  endTrial();
              }
          }
```
or earlier, in the  "const showTarget = () => { ... } " block? 

Apologies if this is a complex question, but if you could give me some pointers it would be very helpful.
